File: src/harness_core/dispatch/orchestration/schemas.py
"""Phase 5 orchestration schemas: workflows, nodes, edges, messages, conflicts, roles."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

# ---------------------------------------------------------------------------
# Schema versions
# ---------------------------------------------------------------------------

WORKFLOW_SCHEMA_VERSION = "workflow_graph.v1"
WORKFLOW_NODE_SCHEMA_VERSION = "workflow_node.v1"
WORKFLOW_EDGE_SCHEMA_VERSION = "workflow_edge.v1"
AGENT_MESSAGE_SCHEMA_VERSION = "agent_message.v1"
CONFLICT_RECORD_SCHEMA_VERSION = "conflict_record.v1"
AGENT_ROLE_SCHEMA_VERSION = "agent_role.v1"

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

WORKFLOW_STATUSES: tuple[str, ...] = (
    "created", "decomposed", "running", "waiting_human",
    "aggregating", "completed", "failed", "cancelled",
)

NODE_STATUSES: tuple[str, ...] = (
    "pending", "ready", "running", "completed", "failed", "cancelled", "waiting_human",
)

EDGE_TYPES: tuple[str, ...] = ("dependency", "data_flow")

MESSAGE_TYPES: tuple[str, ...] = ("task_assign", "result", "conflict", "approval_request", "status_update")

CONFLICT_TYPES: tuple[str, ...] = ("output_conflict", "resource_conflict", "dependency_violation", "budget_overrun")

RESOLUTION_STRATEGIES: tuple[str, ...] = ("latest_wins", "priority_wins", "merge", "human_decides", "skip")

# ---------------------------------------------------------------------------
# Schemas
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class WorkflowNode:
    node_id: str
    workflow_id: str
    task_type: str
    assigned_agent_id: str | None
    status: str = "pending"
    input_refs: tuple[str, ...] = ()
    output_ref: str | None = None
    budget: float = 0.0
    cost_incurred: float = 0.0
    error: str | None = None
    created_at: str = ""
    started_at: str | None = None
    completed_at: str | None = None
    schema_version: str = WORKFLOW_NODE_SCHEMA_VERSION

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "node_id": self.node_id,
            "workflow_id": self.workflow_id,
            "task_type": self.task_type,
            "assigned_agent_id": self.assigned_agent_id,
            "status": self.status,
            "input_refs": list(self.input_refs),
            "output_ref": self.output_ref,
            "budget": self.budget,
            "cost_incurred": self.cost_incurred,
            "error": self.error,
            "created_at": self.created_at,
            "started_at": self.started_at,
            "completed_at": self.completed_at,
        }


@dataclass(frozen=True)
class WorkflowEdge:
    edge_id: str
    from_node_id: str
    to_node_id: str
    edge_type: str = "dependency"
    schema_version: str = WORKFLOW_EDGE_SCHEMA_VERSION

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "edge_id": self.edge_id,
            "from_node_id": self.from_node_id,
            "to_node_id": self.to_node_id,
            "edge_type": self.edge_type,
        }


@dataclass(frozen=True)
class WorkflowGraph:
    workflow_id: str
    dispatch_id: str
    nodes: tuple[WorkflowNode, ...] = ()
    edges: tuple[WorkflowEdge, ...] = ()
    status: str = "created"
    created_at: str = ""
    started_at: str | None = None
    completed_at: str | None = None
    result: dict[str, Any] | None = None
    schema_version: str = WORKFLOW_SCHEMA_VERSION

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "workflow_id": self.workflow_id,
            "dispatch_id": self.dispatch_id,
            "nodes": [n.to_dict() for n in self.nodes],
            "edges": [e.to_dict() for e in self.edges],
            "status": self.status,
            "created_at": self.created_at,
            "started_at": self.started_at,
            "completed_at": self.completed_at,
            "result": self.result,
        }


@dataclass(frozen=True)
class AgentMessage:
    message_id: str
    from_agent_id: str
    to_agent_id: str
    workflow_id: str
    message_type: str
    payload: dict[str, Any] = field(default_factory=dict)
    timestamp: str = ""
    schema_version: str = AGENT_MESSAGE_SCHEMA_VERSION

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "message_id": self.message_id,
            "from_agent_id": self.from_agent_id,
            "to_agent_id": self.to_agent_id,
            "workflow_id": self.workflow_id,
            "message_type": self.message_type,
            "payload": self.payload,
            "timestamp": self.timestamp,
        }


@dataclass(frozen=True)
class ConflictRecord:
    conflict_id: str
    workflow_id: str
    conflict_type: str
    involved_nodes: tuple[str, ...]
    resolution_strategy: str | None = None
    resolution_result: str | None = None
    resolved_at: str | None = None
    schema_version: str = CONFLICT_RECORD_SCHEMA_VERSION

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "conflict_id": self.conflict_id,
            "workflow_id": self.workflow_id,
            "conflict_type": self.conflict_type,
            "involved_nodes": list(self.involved_nodes),
            "resolution_strategy": self.resolution_strategy,
            "resolution_result": self.resolution_result,
            "resolved_at": self.resolved_at,
        }


@dataclass(frozen=True)
class AgentRole:
    role_id: str
    role_name: str
    capabilities: tuple[str, ...] = ()
    max_concurrent_nodes: int = 1
    budget_limit: float = 0.0
    schema_version: str = AGENT_ROLE_SCHEMA_VERSION

    def to_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "role_id": self.role_id,
            "role_name": self.role_name,
            "capabilities": list(self.capabilities),
            "max_concurrent_nodes": self.max_concurrent_nodes,
            "budget_limit": self.budget_limit,
        }

File: engine/src/orchestration/schemas.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;

// Schema versions
pub const WORKFLOW_SCHEMA_VERSION: &str = "workflow_graph.v1";
pub const WORKFLOW_NODE_SCHEMA_VERSION: &str = "workflow_node.v1";
pub const WORKFLOW_EDGE_SCHEMA_VERSION: &str = "workflow_edge.v1";
pub const AGENT_MESSAGE_SCHEMA_VERSION: &str = "agent_message.v1";
pub const AGENT_STATE_SCHEMA_VERSION: &str = "agent_state.v1";
pub const CONFLICT_RECORD_SCHEMA_VERSION: &str = "conflict_record.v1";
pub const AGENT_ROLE_SCHEMA_VERSION: &str = "agent_role.v1";
pub const CHILD_TASK_PROPOSAL_SCHEMA_VERSION: &str = "child_task_proposal.v1";
pub const HANDOFF_REQUEST_SCHEMA_VERSION: &str = "handoff_request.v1";

pub const MAILBOX_STATUSES: &[&str] = &["pending", "read", "acked", "replied", "cancelled"];
pub const PROPOSAL_STATUSES: &[&str] = &["pending", "accepted", "rejected", "cancelled"];
pub const PROPOSAL_TYPES: &[&str] = &[
    "child_task",
    "handoff",
    "review_request",
    "review_verdict",
    "debate_request",
    "debate_position",
    "debate_resolution",
];

pub const AGENT_STATUSES: &[&str] = &["idle", "busy", "blocked", "completed", "failed"];

// Constants
pub const WORKFLOW_STATUSES: &[&str] = &[
    "created",
    "decomposed",
    "running",
    "waiting_human",
    "aggregating",
    "completed",
    "failed",
    "cancelled",
];
pub const NODE_STATUSES: &[&str] = &[
    "pending",
    "ready",
    "running",
    "completed",
    "failed",
    "cancelled",
    "waiting_human",
];
pub const EDGE_TYPES: &[&str] = &["dependency", "data_flow"];
pub const MESSAGE_TYPES: &[&str] = &[
    "task_assign",
    "result",
    "conflict",
    "approval_request",
    "status_update",
];
pub const CONFLICT_TYPES: &[&str] = &[
    "output_conflict",
    "resource_conflict",
    "dependency_violation",
    "budget_overrun",
];
pub const RESOLUTION_STRATEGIES: &[&str] = &[
    "latest_wins",
    "priority_wins",
    "merge",
    "human_decides",
    "skip",
];

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct WorkflowNode {
    pub schema_version: String,
    pub node_id: String,
    pub workflow_id: String,
    pub task_type: String,
    pub assigned_agent_id: Option<String>,
    pub status: String,
    pub input_refs: Vec<String>,
    pub output_ref: Option<String>,
    pub budget: f64,
    pub cost_incurred: f64,
    pub error: Option<String>,
    pub created_at: String,
    pub started_at: Option<String>,
    pub completed_at: Option<String>,
}

impl WorkflowNode {
    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).unwrap_or(Value::Null)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct WorkflowEdge {
    pub schema_version: String,
    pub edge_id: String,
    pub from_node_id: String,
    pub to_node_id: String,
    pub edge_type: String,
}

impl WorkflowEdge {
    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).unwrap_or(Value::Null)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct WorkflowGraph {
    pub schema_version: String,
    pub workflow_id: String,
    pub dispatch_id: String,
    pub nodes: Vec<WorkflowNode>,
    pub edges: Vec<WorkflowEdge>,
    pub status: String,
    pub created_at: String,
    pub updated_at: String,
    pub started_at: Option<String>,
    pub completed_at: Option<String>,
    pub result: Option<Value>,
}

impl WorkflowGraph {
    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).unwrap_or(Value::Null)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AgentMessage {
    pub schema_version: String,
    pub message_id: String,
    pub from_agent_id: String,
    pub to_agent_id: String,
    pub workflow_id: String,
    pub message_type: String,
    pub payload: HashMap<String, Value>,
    pub timestamp: String,
}

impl AgentMessage {
    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).unwrap_or(Value::Null)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AgentState {
    pub schema_version: String,
    pub agent_id: String,
    pub run_id: String,
    pub role: String,
    pub capability_profile: Vec<String>,
    pub objective: Option<String>,
    pub status: String,
    pub scratchpad_summary: Option<String>,
    pub redaction_filter: Option<String>,
    pub metadata: HashMap<String, Value>,
    pub created_at: String,
    pub updated_at: String,
}

impl AgentState {
    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).unwrap_or(Value::Null)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MailboxMessage {
    pub schema_version: String,
    pub message_id: String,
    pub correlation_id: Option<String>,
    pub from_agent_id: String,
    pub to_agent_id: String,
    pub run_id: Option<String>,
    pub node_id: Option<String>,
    pub message_type: String,
    pub status: String,
    pub body: Option<String>,
    pub body_summary: Option<String>,
    pub redaction_status: String,
    pub created_at: String,
    pub read_at: Option<String>,
    pub ack_at: Option<String>,
    pub reply_to_message_id: Option<String>,
    pub metadata: HashMap<String, Value>,
}

impl MailboxMessage {
    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).unwrap_or(Value::Null)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SendMessageRequest {
    pub from_agent_id: String,
    pub to_agent_id: String,
    pub run_id: Option<String>,
    pub node_id: Option<String>,
    pub message_type: String,
    pub body: Option<String>,
    pub correlation_id: Option<String>,
    pub reply_to_message_id: Option<String>,
    pub metadata: Option<HashMap<String, Value>>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ListMailboxQuery {
    pub agent_id: Option<String>,
    pub run_id: Option<String>,
    pub node_id: Option<String>,
    pub status: Option<String>,
    pub limit: Option<i64>,
    pub offset: Option<i64>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ConflictRecord {
    pub schema_version: String,
    pub conflict_id: String,
    pub workflow_id: String,
    pub conflict_type: String,
    pub involved_nodes: Vec<String>,
    pub resolution_strategy: Option<String>,
    pub resolution_result: Option<String>,
    pub resolved_at: Option<String>,
}

impl ConflictRecord {
    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).unwrap_or(Value::Null)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AgentRole {
    pub schema_version: String,
    pub role_id: String,
    pub role_name: String,
    pub capabilities: Vec<String>,
    pub max_concurrent_nodes: usize,
    pub budget_limit: f64,
}

impl AgentRole {
    pub fn to_dict(&self) -> Value {
        serde_json::to_value(self).unwrap_or(Value::Null)
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ChildTaskProposal {
    pub schema_version: String,
    pub correlation_id: String,
    pub objective: String,
    pub context_summary: String,
    pub proposed_node_id: Option<String>,
    pub proposed_edge_id: Option<String>,
    pub parent_node_id: String,
    pub run_id: String,
    pub agent_id: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct HandoffRequest {
    pub schema_version: String,
    pub correlation_id: String,
    pub objective: String,
    pub context_summary: String,
    pub target_agent_id: String,
    pub source_agent_id: String,
    pub run_id: String,
    pub node_id: String,
}

// ── AR-5: Bounded review/debate primitives ────────────────────────────────────

pub const MAX_DEBATE_PARTICIPANTS: usize = 8;
pub const MAX_DEBATE_ROUNDS: usize = 10;
pub const MAX_REVIEW_DEBATE_TEXT_BYTES: usize = 4096;

pub const REVIEW_VERDICTS: &[&str] = &["accepted", "rejected"];
pub const DEBATE_POSITION_STATUSES: &[&str] = &["pending", "accepted", "rejected"];

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ReviewRequest {
    pub schema_version: String,
    pub correlation_id: String,
    pub subject_summary: String,
    pub rationale_summary: String,
    pub target_agent_id: String,
    pub run_id: String,
    pub node_id: String,
    pub blocking: bool,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ReviewVerdict {
    pub schema_version: String,
    pub correlation_id: String,
    pub review_request_id: String,
    pub verdict: String,
    pub rationale_summary: String,
    pub run_id: String,
    pub node_id: String,
    pub blocking: bool,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DebateRequest {
    pub schema_version: String,
    pub correlation_id: String,
    pub subject_summary: String,
    pub participant_agent_ids: Vec<String>,
    pub max_rounds: usize,
    pub run_id: String,
    pub node_id: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DebatePosition {
    pub schema_version: String,
    pub correlation_id: String,
    pub debate_id: String,
    pub position: String,
    pub rationale_summary: String,
    pub run_id: String,
    pub node_id: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DebateResolution {
    pub schema_version: String,
    pub correlation_id: String,
    pub debate_id: String,
    pub resolution: String,
    pub winning_position: Option<String>,
    pub unresolved_risks: Option<String>,
    pub run_id: String,
    pub node_id: String,
}
